File: dwcodebook/inst/perltools/parsecontents.pl
#################################################################################################################################
## Author: BJ Harshfield,  Ariana Huebner ##
## Date: 11-19-2012 ##
## Program Name: parsecontents.pl ##
## Version: 1 ##
## Project: R codebook ##
## Description: This is a simple script parse sas proc contents files for use with createcb()
## Notes: Need to test this on more files             
##              
##              
##		
################################################################################################################################

#open in log
$ifile=$ARGV[0];
chomp $ifile;
#$ifile="contents.saslog";

#open output
$ofile=$ARGV[1];
chomp $ofile;
#$ofile="output.csv";

open(IN,"$ifile");
open(OUT,">$ofile");
print OUT "var,label,form\n";

$on=0;
while(<IN>){

	
	if($_=~/The SAS System/){$on=0};
	if($_=~/\w/ & ($on==1)){
		#get char vector
		@temp=split//,$_;
		
		#get the var string
		$var="";
		for($x=6;$x<=36;$x++){$var=$var.$temp[$x]}
		$var=~s/\s+$//;		
		print OUT $var;
		print OUT ",";		

		#get the label string
		$lab="";
		for($x=48;$x<=$#temp;$x++){$lab=$lab.$temp[$x]}
		$lab=~s/\s+$//;	
		print OUT "\"$lab\"";
		print OUT ",\n";		
		
	}	
	if($_=~/ #  Variable                       Type  Len  Label/){$on=1};	
	

}

close IN;
close OUT;

File: dwcodebook/R/parsecontents.R
#################################################################################################################################
## Author: BJ Harshfield,  Ariana Huebner ##
## Date: 11-16-2012 ##
## Program Name: parsesascontents ##
## Version: 1 ##
## Project: R codebook ##
## Description: This is a simple script to create a variable description data frame from a sas proc contents file.
##              It is a wrapper for a simple perl script that does the extraction.
##             
##              
##		
################################################################################################################################


#parse parameters from perl script
#$ifile=$ARGV[0];
#$ofile=$ARGV[1];


parsesascontents<-function(ifile,ofile="temp.csv",addallvars=FALSE){
	#send the command to generate the csv from the contents file
	cmd=paste("perl ",system.file(package="dwcodebook","perltools/parsecontents.pl"), " \"",ifile,"\" \"",ofile,"\"",sep="")
	#cat(cmd,"\n");	
	system(cmd)
	#load in the file
	temp=read.csv(ofile,stringsAsFactors=FALSE)
	if(addallvars){temp$allvars=TRUE}
	temp	
	
}



